### PR TITLE
[Ingest Manager] Fix failing installation on windows 7

### LINKED ISSUE #24327: [Agent] Installed Windows 7 Agent shows no logs in Kibana (data is in Discover though) and shows host log error: agent events contains events with different agent id from currently authorized agent
This is a 7.12 BC3 stack deploy to cloud with the BC3 Agent on a Win7 VM

Agent shows as 'healthy' but no logs show in the Agent details, presumably due to the mis-match somewhere of the Agent ID.  With recent changes on Install in Beats (and none on Kibana side that seem to relate) I assume it is a Beats Agent bug.

The first error in the log ends up being repeated, over an over, stacking up the # of actions that are not dispatched, the first was:

`{"log.level":"error","@timestamp":"2021-03-03T14:03:54.788-0500","log.origin":{"file.name":"application/fleet_gateway.go","file.line":185},"message":"failed to dispatch actions, error: acknowledge 1 actions '[action_id: 59c47640-7c44-11eb-b1f6-1755d2a3f765, type: POLICY_CHANGE]' for elastic-agent 'cbad3b28-720c-4bdf-8d58-988047ae540e' failed: Status code: 400, Kibana returned an error: Bad Request, message: agent events contains events with different agent id from currently authorized agent","ecs.version":"1.6.0"}
`

Full Logs zip attached along with state.yml (changed to .txt for git inclusion)
[state.yml.txt](https://github.com/elastic/beats/files/6078656/state.yml.txt)
[7.12-bc3-win-7-logs.zip](https://github.com/elastic/beats/files/6078657/7.12-bc3-win-7-logs.zip)

The Agent does seem to be sending data to ES and functioning, perhaps with just some default set policy.

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

Issue described here: #24327 
there was a race between enrollment process and restarting service, FS playing part as well. 
The thing was that when agent was restarted only on windows 7 it loaded standalone ID, even though it was already replaced by enrollment process. 
Then when agent retrieved hosts from fleet it even overwrote updated ID with stale one. 

This fix adds a lock which prevents simultaneous write in between these two processes and a forced Reload in case of fleet managed agent later in the cycle. 
Another thing is FSync after file rotation which was missing for windows. 

These seems to fix the issue, tested on win 7 VM on cloud

## Why is it important?

Fixes #24327

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: x-pack/elastic-agent/pkg/agent/application/enroll_cmd.go [hunk 1]
```diff
@@ -557,3 +553,31 @@ func getAppFromStatus(status *client.AgentStatus, name string) *client.Applicati
...
 }
+
+func safelyStoreAgentInfo(s saver, reader io.Reader) error {
+	err := storeAgentInfo(s, reader)
+	signal := make(chan struct{})
+	backExp := backoff.NewExpBackoff(signal, 200*time.Millisecond, 15*time.Second)
+
+	for err != nil {
+		backExp.Wait()
+		err = storeAgentInfo(s, reader)
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/agent/application/enroll_cmd.go:564]: What if this always fails? How does this not loop forever?

**Review by blakerouse — CHANGES_REQUESTED** [review]
I think an infinite loop is present.

File: x-pack/elastic-agent/pkg/agent/application/enroll_cmd.go (hunk 1 above)
Comment (michalpristas) [reply x-pack/elastic-agent/pkg/agent/application/enroll_cmd.go:566]: yeah copy paste i rewrote that part

**Review by blakerouse — APPROVED** [review]
Retry logic looks correct now.

**Comment by dikshachauhan-qasource** [comment]
Hi @EricDavisX 

Today we have validated the above fixes on upgraded Kibana from 7.10.2 to 7.12 and found the issue as fixed.

Installed agent with only system integration in Agent policy:

Observations:
- Agent is in healthy state.
- Activity logs present
- logs on discover tab are visible.
- Metricbeat and filebeat are running.

image

Build details:
```
BUILD 39134
COMMIT 08417cbd6c15e4c866651a7dcdfeded58845206d
Artifact link: https://staging.elastic.co/7.12.0-96914cb5/summary-7.12.0.html
```

Thanks
QAS

**Pull request merged**